include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,216 @@
+	{
+		time_point added;
+		tcp::endpoint addr;
+		bool seed;
+	};
+
+	// this is a group. It contains a set of group members
+	struct torrent_entry
+	{
+		std::string name;
+		std::set<peer_entry> peers;
+	};
+
+	struct dht_immutable_item
+	{
+		dht_immutable_item() : value(0), size(0) {}
+		// malloced space for the actual value
+		char* value;
+		// size of malloced space pointed to by value
+		int size;
+	};
+
+	struct ed25519_public_key { char bytes[item_pk_len]; };
+
+	struct dht_mutable_item : dht_immutable_item
+	{
+		char sig[item_sig_len];
+		boost::uint64_t seq;
+		ed25519_public_key key;
+		char* salt;
+		int salt_size;
+	};
+
+	struct TORRENT_EXPORT dht_storage_interface
+	{
+#ifndef TORRENT_NO_DEPRECATE
+		virtual size_t num_torrents() const = 0;
+		virtual size_t num_peers() const = 0;
+#endif
+
+		// returns true if the entry is found and it is returned in the
+		// out parameter entry
+		virtual bool lookup_peers(sha1_hash const& info_hash
+			, torrent_entry& entry) const = 0;
+		// this method is named announce_peer for consistency with the
+		// upper layers, but has nothing to do with networking. It's only
+		// responsibility is store the peer in such a way that it's returned
+		// in the torrent_entry with the lookup_peers.
+		virtual void announce_peer(sha1_hash const& info_hash
+			, tcp::endpoint const& endp
+			, std::string const& name, bool seed) = 0;
+
+		// returns true if the item is found and it is returned in the
+		// out parameter item
+		virtual bool get_immutable_item(sha1_hash const& target
+			, dht_immutable_item& item) const = 0;
+		virtual void put_immutable_item(sha1_hash const& target
+			, dht_immutable_item const& to_add, address addr) = 0;
+
+		// returns true if the item is found and it is returned in the
+		// out parameter item
+		virtual bool get_mutable_item(sha1_hash const& target
+			, dht_mutable_item& item) const = 0;
+		virtual void put_mutable_item(sha1_hash const& target
+			, dht_mutable_item const& to_add, address addr) = 0;
+
+		virtual void tick() = 0;
+	};
+
+	// NON-PUBLIC TYPES
+	// the following types are private, pending review to move them
+	// to a better place
+
+	struct dht_popularity_item
+	{
+		dht_popularity_item() : num_announcers(0) {}
+		// this counts the number of IPs we have seen
+		// announcing this item, this is used to determine
+		// popularity if we reach the limit of items to store
+		bloom_filter<128> ips;
+		// the last time we heard about this
+		time_point last_seen;
+		// number of IPs in the bloom filter
+		int num_announcers;
+	};
+
+	struct dht_immutable_entry
+	{
+		dht_immutable_entry(const dht_immutable_item& item) : item(item) {}
+
+		dht_immutable_item item;
+		dht_popularity_item popularity;
+	};
+
+	struct dht_mutable_entry
+	{
+		dht_mutable_entry(const dht_mutable_item& item) : item(item) {}
+
+		dht_mutable_item item;
+		dht_popularity_item popularity;
+	};
+
+	class dht_default_storage : public dht_storage_interface, boost::noncopyable
```

Comment:
once the storage in the node class is a dht_storage_interface, the concrete type of the default storage can be hidden away inside dht_storage.cpp (or possibly some other .cpp file)